### ISSUE TITLE
Dropdown-panel size issue

### ISSUE THREAD

**Issue by israel-hdez** [patternfly/patternfly-react#3156]
When using the generic Dropdown panel pattern (https://www.patternfly.org/v4/documentation/react/components/dropdown/#dropdown-panel), it is possible to place arbitrary content in the panel:

![image](https://user-images.githubusercontent.com/23639005/66958957-05073b80-f02f-11e9-89fe-6a6855a1b2a0.png)

At some point, the generated markup of Dropdown changed to remove `ul` DOM nodes, which is fine. However, there is an issue that is causing the panel of the dropdown to limit it's size to the width of the dropdown. Previously, the width of the panel increased as needed (similarly the other patterns). This generates issues when the panel's content needs a larger width:

![image](https://user-images.githubusercontent.com/23639005/66959359-cc1b9680-f02f-11e9-811d-33cae204747c.png)

For now, to workaround the issue I'm adding an additional layer with a `className="pf-c-dropdown__menu-item"` like the following (which I formed by looking at the markup of older react-core versions):

```tsx
<Dropdown
    onSelect={this.onSelect}
    toggle={<DropdownToggle onToggle={this.onToggle}>Expanded Dropdown</DropdownToggle>}
    isOpen={isOpen}
    >
    <div className="pf-c-dropdown__menu-item">
        [Panel contents here]
    </div>
</Dropdown>
```

and this workaround fixes the sizing issue:

![image](https://user-images.githubusercontent.com/23639005/66959567-40eed080-f030-11e9-9747-94ecd82a1b8d.png)

But I think the "generic Dropdown" pattern should generate the required markup to avoid this workaround and to behave the same as the other patterns.

**Comment by tlabaj** [comment]
cc @mcoker @jgiardino

**Comment by mcoker** [comment]
Hi @israel-hdez in this case, it would still be advisable to use `.pf-c-dropdown__menu-item` so that you get the default styling provided by the dropdown component. I would use that in combination with `label.pf-c-check` for your checkboxes. I put together an example here, though I realized in doing so, I think we have a bug where `.pf-c-dropdown__menu-item` is set to `display: block` by default, and `.pf-c-check` needs to be `display: grid`, so you can see where I added that back in at the top of the example.

https://codesandbox.io/embed/reverent-shockley-gejui

**Comment by mcoker** [comment]
@israel-hdez it looks like this use case might be better as an [<OptionsMenu>](https://patternfly-react.surge.sh/patternfly-4/documentation/react/components/optionsmenu) or [<Select>](https://patternfly-react.surge.sh/patternfly-4/documentation/react/components/select) component? The dropdown should only handle links or actions.

Even with a custom menu, it's still advised to use the `.pf-c-[select/options-menu/dropdown]--menu-item` class on your items so that you retain the PF styles (layout, spacing, hover/focus states). You can see an example of your general layout using the select component here - https://www.patternfly.org/v4/documentation/react/components/select/#grouped-checkbox-select-input

**Comment by israel-hdez** [comment]
@mcoker The **Select** component looks suitable for the exposed case. I'll use it.

However, there is a very particular dropdown where none of your suggested components look suitable. It looks like this:

![image](https://user-images.githubusercontent.com/23639005/67420193-c1af4e80-f593-11e9-8b56-c4ce687eed7a.png)

It looks very similar, but the main difference is that there is a 2nd level of checkboxes which are shown/hidden when the parent is checked/unchecked (because the nested items are retrieved on-demand).

But, honestly, regardless of my use-cases, I think the dropdown should be fixed because it offers that free-form panel which is very useful for custom cases.

Anyway, thanks for your suggestion! I'll use your suggested components where I can.

**Comment by mcoker** [comment]
@mcarrano what do you think about the nested structure above? Is that something we should consider for enhancement?

> But, honestly, regardless of my use-cases, I think the dropdown should be fixed because it offers that free-form panel which is very useful for custom cases.

We just added support for this in the select component as well in patternfly-core (https://github.com/patternfly/patternfly-next/pull/2337) and have a follow up issue to show an example of it in patternfly-react (https://github.com/patternfly/patternfly-react/issues/3193)

**Comment by mcarrano** [comment]
I agree with @israel-hdez that the width of the dropdown panel should not be constrained by the width of the field. As for supporting nested checkboxes inside of a Checkbox Select, that's something we could add if there is a general need for it. Otherwise I think it's OK to accomplish by adding to the free-form panel as being done here.

**Comment by mcoker** [comment]
Thanks @mcarrano 

@israel-hdez regarding the word wrapping in the custom menu, as I mentioned it's still expected that you would use `.pf-c-dropdown__menu-item` (or `.pf-c-select__menu-item`, etc) on the clickable items in the menu if you can. That defines a number of styles that we define as defaults for the component - https://github.com/patternfly/patternfly-next/blob/289f99b4272619a0b2618610b4d1c8d3a3861cd7/src/patternfly/components/Dropdown/dropdown.scss#L329-L359. One of which is `white-space: nowrap;`, which disables line wrapping.  If you opt out of using the `__menu-item` class on selectable items, it's expected that you will add your own custom styles to those elements, including disabling the line wrapping.

**Comment by israel-hdez** [comment]
@mcoker Then, probably, a more complete example is needed in the react docs at https://www.patternfly.org/v4/documentation/react/components/dropdown/#dropdown-panel

The current example in the docs shows this simple markup and it doesn't give any clue that the free-form panel should have `.pf-c-dropdown__menu-item` styles (or perhaps it's me who can't find that advice in the docs):

```jsx
<Dropdown
    onSelect={this.onSelect}
    toggle={<DropdownToggle onToggle={this.onToggle}>Expanded Dropdown</DropdownToggle>}
    isOpen={isOpen}
>
    [Panel contents here]
</Dropdown>
```

(note that I'm specifically talking about the PF React library, where manually adding styles sounds counterintuitive. In the HTML version it may be expected because in HTML we explicitly write all the markup).

BTW, I'm moving to using the `Select` control where possible. Is there any way to get rid of the selection count? I think that the selection count is to show how many *filters* are selected. But in my use-case I'm listing bare *options* (not filters), thus displaying the selection count is not useful (it's just wasting screen space). I'm talking about this:

![image](https://user-images.githubusercontent.com/23639005/67805924-8ad7ad80-fa57-11e9-8c5d-90002f45cc0b.png)

**Comment by tlabaj** [comment]
@israel-hdez we can not hide the selection count badge at the moment.  
@mcarrano we can add the functionality to hide the badge in the dropdown?  Do you have any concerns about this from a design perspective?

**Comment by mcoker** [comment]
@israel-hdez my apologies for the delay in responding, I lost track of this issue in my backlog. 

> The current example in the docs shows this simple markup and it doesn't give any clue that the free-form panel should have `.pf-c-dropdown__menu-item` styles (or perhaps it's me who can't find that advice in the docs).

This would just be my recommendation if you were using a custom menu panel and wanted to retain the styles associated with the existing dropdown items (`white-space: nowrap;` and the hover/focus styles, for example) - it isn't a requirement by any means. I would consider it an uncommon use case since the most likely usage of the dropdown panel example is for a custom list of items where the existing dropdown item styling isn't preferred. I suggested it based on the design you provided, which seems like it should retain the menu styles provided by the component. However the select seems much better suited for your use, and it gives you the functionality and styling you're looking for by default (except for the badge issue @tlabaj responded to).

Writing the dropdown item class names manually in the code would be the only way to retain the menu item styles in the dropdown panel example currently, which I agree isn't ideal, however I've yet to come across a use case for needing to do this, where another solution wasn't available (using the select component, for example). If we encounter the need to do this in the future, we can provide an easier way to do this than currently exists along with docs and examples.

**Comment by mcarrano** [comment]
@israel-hdez @tlabaj I'm not opposed to making the badge that counts the number of items selected optional, however I'm wondering if the Options Menu is a better choice here (https://www.patternfly.org/v4/documentation/react/components/optionsmenu) It was designed for this type of use case. The only thing I was unclear about is whether it will support multi-selections like this, but I think it should. Would that work here?

**Comment by israel-hdez** [comment]
Sorry for replying slowly.

@mcarrano Unfortunately, I prefer the explicit checkboxes because it clearly gives to the user the idea that items can be checked/unchecked. The Options Menu may confuse the user if nothing is checked. Additionally, the Options Menu seems to not provide the headers of each group, which are somewhat important in our use-case to properly give context.

**Comment by mcarrano** [comment]
@israel-hdez What you are asking for makes sense. Just wanted to offer that as a possible alternative before we committed to modify the checkbox select. Can you summarize again exactly what you are requesting as there seemed to be multiple requests discussed in the comments. Some may be easier to address than others. Also giving us some idea of what your project timeline is will help prioritize this.

**Comment by israel-hdez** [comment]
@mcarrano We are going feature freeze this month. But nothing here is critical, so no need to hurry :)

Summarizing, I think the only resulting request is the ability to hide the badge that counts the number of checked items on the Select control.

The original issue of the sizing of the Dropdown control when using the "generic" pattern is solved by adding the `.pf-c-dropdown__menu-item` as @mcoker suggested. I suggested to have a note about this in the docs, but as @mcoker says, my use-case could be too specific and no doc note may be needed.

**Comment by mcarrano** [comment]
@israel-hdez I've opened a new issue here to track the request to make the badge optional: https://github.com/patternfly/patternfly-next/issues/2519 Will close this one out,
